psi4/driver/p4util/prop_util.py

```diff
@@ -0,0 +1,140 @@
+#
+# @BEGIN LICENSE
+#
+# Psi4: an open-source quantum chemistry software package
+#
+# Copyright (c) 2007-2021 The Psi4 Developers.
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This file is part of Psi4.
+#
+# Psi4 is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# Psi4 is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+"""Module with property-related helper functions."""
+
+import qcelemental as qcel
+
+import psi4
+from psi4 import core
+from . import optproc
+
+
+def free_atom_volumes(wfn, **kwargs):
+    """ 
+
+    """
+
+    # the level of theory
+    module = wfn.module()
+
+    # print level
+    print_level = core.get_global_option("PRINT")
+
+    # if we're doing dft, grab the functional
+    if module.lower() == 'scf':
+        module = wfn.functional().name()
+    # list of reference number of unpaired electrons.
+    # Note that this is not the same as the 
+    # total spin of the ground state atom
+    reference_S = [ 0,
+                    1,                                                                                           0,
+                    1, 0,                                                                         1, 2, 3, 2, 1, 0,
+                    1, 0,                                                                         1, 2, 3, 2, 1, 0,
+                    1, 0,                                           1, 2, 3, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0,
+                    1, 0,                                           1, 2, 5, 6, 5, 4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0,
+                    1, 0, 1, 0, 3, 4, 5, 6, 7, 8, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0 ]
+
+
+    # the parent molecule and reference type
+    mol = wfn.molecule()
+    user_ref = core.get_global_option('REFERENCE')
+
+    # Get unique atoms by input symbol,
+    # Be to handle different basis sets
+    atom_ids = {}
+    for atom in range(mol.natom()):
+        symbol = mol.symbol(atom)
+        Z = int(mol.Z(atom))
+        basis = mol.basis_on_atom(atom)
+    
+        for a, info in atom_ids.items():
```

Comment:
If I understand this correctly, you only want unique atoms (as defined by symbol and basis), and this `for` loop is supposed to filter out duplicates. This `for` loop doesn't accomplish that, because it only `continue`s the inner loop, not the outer loop.

But in this case, why are you using a `dict` structure for `atom_ids` at all? You don't do anything with the labels. I'd replace your `dict` of `dict`s with a `set` of `tuple` or `collections.namedtuple`, as a clearer data structure. Even a `list` of `dict`s would be cleaner. The advantage of a `set` is that it filters duplicates for you.

Reply:
Using `set` is a good suggestion, thanks, and a definite blunder on my part for not filtering duplicates on the most recent commit.